### PR TITLE
♿ 🐛 Show Carousel control button when mouse is detected

### LINKED ISSUE #29560: amp-carousel: No indication of focus (hidden controls)
## WCAG Level
Level AA

## Priority
High

## Type
Navigation

## Pages/screens/Components affected
* [amp-carousel](https://amp.dev/documentation/components/amp-carousel/)

## Description
Keyboard users rely on the browser's focus outline to keep track of where they are on web pages. The currently selected form input, button, or hyperlink is the one said to have the focus.

Most browsers use a faint dotted outline to indicate the focus to the user, though some browsers like Chrome or Safari use an enhanced focus outline, consisting of a coloured border around the selected link. It is also possible for developers to provide custom focus indications within the site's cascading style sheets (CSS).

By default (unless the optional `controls` attribute has been set) `<amp-carousel>` hides the previous/next buttons after a user has interacted with the page using touch. The buttons are still focusable, and can still be reached by keyboard users, but they remain completely invisible. These controls then fail WCAG 2.1 SC 2.4.7 Focus Visible, as they provide no visible focus indication (on top of fundamentally being invisible themselves).

Screenshot of the carousel in small-screen touch device emulation, showing that the buttons are visually hidden, but still present in the page and receiving focus

These controls are then also completely unusable for mouse users (as they're visually hidden, and their mouse interactions are suppressed via CSS).

```
amp-carousel[i-amphtml-carousel-hide-buttons] .amp-carousel-button-next, amp-carousel[i-amphtml-carousel-hide-buttons] .amp-carousel-button-prev {
    opacity: 0;
    pointer-events: none;
    visibility: visible!important;
}
```

Note that this aspect (inability to use mouse or keyboard after touch interaction) falls outside of WCAG 2.1 Level A/AA (which this evaluation was based on), but does fail WCAG 2.1 Success Criterion 2.5.6 Concurrent Input Mechanisms (Level AAA).

## User impact
The default behaviour of carousels does not take into account users who may be using a touchscreen in conjunction with other input methods, such as a keyboard or mouse. They may be on a laptop/desktop with a touchscreen (e.g. Surface laptop), or use a "touch" device such as a smartphone or tablet, but have a paired bluetooth mouse and/or keyboard. For users who occasionally use the touchscreen, but then switch to any of their other input mechanisms, the carousels will become unusable unless they, after the first touchscreen interaction, consistently only use touch to operate the carousels.

Once the controls are hidden, but receive keyboard focus, the problem for a keyboard user will be that without a visible focus indicator it can be difficult or impossible for these users to know which control (whether it be a link, button, or something else) currently has focus and therefore what action will take place if they press the `Enter`, `Space`, or other keys.

## Required solution
While it may be appropriate to visually hide the previous/next buttons when the user interacts with the page using a touchscreen, this should be reverted - and the buttons be made visible - as soon as the user switches to using a mouse or keyboard again. Once the buttons are visible again, this will also resolve the issue of visible focus indication (though refer to the separate issue about focus indication in Firefox being insufficient).

## Test procedure(s)
Use these steps to confirm that the solution has been correctly applied to issues identified within the test sample, and to test the rest of the website for instances of the same issue:
1. Interact with the page/carousel using a touchscreen. Wait for the previous/next buttons to disappear/
2. Switch to the keyboard and navigate the page (`Tab`/`Shift`+`Tab`).
3. Verify that the previous/next controls become visible again, and that focus does not "disappear" (i.e. that the focus doesn't end up on invisible elements)

## Related standards
* [WCAG 2.1 Success Criterion 2.4.7 Focus Visible (Level AA)](https://www.w3.org/TR/WCAG21/#focus-visible)
* [WCAG 2.1 Success Criterion 2.5.6 Concurrent Input Mechanisms (Level AAA)](https://www.w3.org/TR/WCAG21/#concurrent-input-mechanisms)

## More information
* [W3C - Failure of Success Criterion 2.4.7 due to styling element outlines and borders in a way that removes or renders non-visible the visual focus indicator](https://www.w3.org/WAI/WCAG21/Techniques/failures/F78.html)

## Test data
Test date: July 2020

### PR DESCRIPTION
**Pull request by zhouyx**
Closes #29560

### CODE REVIEW

**Comment by zhouyx** [comment]
I couldn't find a related test. Open to test ideas.

File: extensions/amp-carousel/0.2/amp-carousel.js [hunk 1]
```diff
@@ -193,6 +200,21 @@ class AmpCarousel extends AMP.BaseElement {
     // Need to wait for slides to exist first.
     this.carousel_.goToSlide(Number(this.element.getAttribute('slide') || '0'));
     // Signal for runtime to check children for layout.
+
+    if (
+      isAmp4Email(this.element.ownerDocument) ||
```
Comment (caroqliu) [inline extensions/amp-carousel/0.2/amp-carousel.js:205]: `amp-carousel:0.2` is not allowed in email yet, so this check is not necessary.

File: extensions/amp-carousel/0.2/amp-carousel.js [hunk 2]
```diff
@@ -115,6 +116,12 @@ class AmpCarousel extends AMP.BaseElement {
 
     /** @private {?ChildLayoutManager} */
     this.childLayoutManager_ = null;
+
+    /**
+     * Whether to show control buttons
+     * @private {boolean}
+     */
+    this.alwaysShowControls_ = false;
```
Comment (caroqliu) [inline extensions/amp-carousel/0.2/amp-carousel.js:124]: suggestion: We can name this `showControls` instead of `alwaysShowControls`, since it can change

**Review by caroqliu — APPROVED** [review]
Thanks for making this contribution, LGTM!

For testing, touch is not yet supported in e2e -- [this thread](28150) also discusses this, though I'm not sure this counts as a unique use case since this and the previous conversation came up from testing carousel's hidden controls feature. I'm not sure if visual-diff tests support this, but that would also be a good route for confirming whether or not the buttons are there.

In the mean time, a unit test should be sufficient. I'd recommend looking at `test/unit/test-input.js` for firing the `onMouseDetected` states and checking the classes accordingly.  

For manual testing, there is also the test procedures outlined in the original issue. :)

File: extensions/amp-carousel/0.2/amp-carousel.js (hunk 1 above)
Comment (zhouyx) [reply extensions/amp-carousel/0.2/amp-carousel.js]: Ah! that explains, I was wonder why we didn't have the check here : )

File: extensions/amp-carousel/0.2/amp-carousel.js (hunk 2 above)
Comment (zhouyx) [reply extensions/amp-carousel/0.2/amp-carousel.js]: done

**Review by estherkim — APPROVED** [review]
✅ bundle size

**Comment by zhouyx** [comment]
The unit test requires me to mock too much events and is not very helpful imho.  If possible I agree visual diff test would be helpful, but it would still require us to fake the mousemove and also touch event.  Tested manually

**Pull request merged**